src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -140,6 +146,11 @@ public void Complete()
             }
         }
 
+        public void PromiseNotToEnqueue()
```

Comment:
The suggestion was in support of fixing the bug, although I agree that `Complete` and `Cancel` would involve more reworking.

`Complete` makes explicit which caller should be completing: `CSharpCompilation.CompleteTree`. The cases where `TryComplete` is used today are really attempts to `Cancel` instead. Perhaps `TryComplete` should be renamed to `Cancel`.

Note there are indirect uses of `TryComplete` through calls to `DiagnosticQueue.TryComplete` as well. It feels like those should explicit there too. All but one of those calls should be called `Cancel` and the other should be `Complete`.